In [13]:
import os
import pandas as pd
import numpy as np
import math
import cmath
import matplotlib.pyplot as plt
import seaborn as sns

# Data

In [14]:
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 300)

In [15]:
df = pd.read_csv("../data/jetblue_df.csv")

In [16]:
df.head()

Unnamed: 0                             legId  searchDate  flightDate  \
0           5  79eda9f841e226a1e2121d74211e595c  2022-04-16  2022-04-17   
1          29  24c07cf38de8a7a0ed2d7709168042a8  2022-04-16  2022-04-17   
2          57  a291a2b5ac19223c3407cfec58d1fc24  2022-04-16  2022-04-17   
3         146  d53bcf9e8b6523fa2de9420d7cb22196  2022-04-16  2022-04-17   
4         209  5bb138f39804320ece767e747608a50e  2022-04-16  2022-04-17   

  startingAirport destinationAirport fareBasisCode travelDuration  \
0             ATL                BOS      VH0AUEL1        PT2H38M   
1             ATL                BOS      QH0KUEL1        PT2H44M   
2             ATL                CLT      VH0AUEL1        PT8H48M   
3             ATL                DTW      RH0AUEL1         PT5H8M   
4             ATL                EWR      VC0QBEL1         PT2H9M   

   elapsedDays  isBasicEconomy  isRefundable  isNonStop  baseFare  totalFare  \
0            0           False         False       True    217.67     248.60   
1            0           False         False       True    316.73     355.09   
2            0           False         False      False    482.79     542.60   
3            0           False         False      False    269.78     313.61   
4            0           False         False       True    171.16     198.60   

   seatsRemaining  totalTravelDistance segmentsDepartureTimeEpochSeconds  \
0               7                947.0                        1650206700   
1               7                947.0                        1650233220   
2               7               1675.0            1650206700||1650229500   
3               7               1247.0            1650235800||1650246600   
4               7                762.0                        1650195000   

                            segmentsDepartureTimeRaw  \
0                      2022-04-17T10:45:00.000-04:00   
1                      2022-04-17T18:07:00.000-04:00   
2  2022-04-17T10:45:00.000-04:00||2022-04-17T17:0...   
3  2022-04-17T18:50:00.000-04:00||2022-04-17T21:5...   
4                      2022-04-17T07:30:00.000-04:00   

  segmentsArrivalTimeEpochSeconds  \
0                      1650216180   
1                      1650243060   
2          1650216180||1650238380   
3          1650244380||1650254280   
4                      1650202740   

                              segmentsArrivalTimeRaw  \
0                      2022-04-17T13:23:00.000-04:00   
1                      2022-04-17T20:51:00.000-04:00   
2  2022-04-17T13:23:00.000-04:00||2022-04-17T19:3...   
3  2022-04-17T21:13:00.000-04:00||2022-04-17T23:5...   
4                      2022-04-17T09:39:00.000-04:00   

  segmentsArrivalAirportCode segmentsDepartureAirportCode  \
0                        BOS                          ATL   
1                        BOS                          ATL   
2                   BOS||CLT                     ATL||BOS   
3                   JFK||DTW                     ATL||JFK   
4                        EWR                          ATL   

                segmentsAirlineName segmentsAirlineCode  \
0                   JetBlue Airways                  B6   
1                   JetBlue Airways                  B6   
2  JetBlue Airways||JetBlue Airways              B6||B6   
3  JetBlue Airways||JetBlue Airways              B6||B6   
4                   JetBlue Airways                  B6   

  segmentsEquipmentDescription segmentsDurationInSeconds segmentsDistance  \
0                          NaN                      9480              947   
1                          NaN                      9840              947   
2                ||Embraer 190                9480||8880         947||728   
3     Embraer 190||Embraer 190                8580||7680         762||485   
4                  Embraer 190                      7740              762   

  segmentsCabinCode  
0             coach  
1             coach  
2      coach||coach  
3      coach||coach  
4      

In [17]:
sorted(df.columns)

['Unnamed: 0',
 'baseFare',
 'destinationAirport',
 'elapsedDays',
 'fareBasisCode',
 'flightDate',
 'isBasicEconomy',
 'isNonStop',
 'isRefundable',
 'legId',
 'searchDate',
 'seatsRemaining',
 'segmentsAirlineCode',
 'segmentsAirlineName',
 'segmentsArrivalAirportCode',
 'segmentsArrivalTimeEpochSeconds',
 'segmentsArrivalTimeRaw',
 'segmentsCabinCode',
 'segmentsDepartureAirportCode',
 'segmentsDepartureTimeEpochSeconds',
 'segmentsDepartureTimeRaw',
 'segmentsDistance',
 'segmentsDurationInSeconds',
 'segmentsEquipmentDescription',
 'startingAirport',
 'totalFare',
 'totalTravelDistance',
 'travelDuration']

In [18]:
df.shape

(6824440, 28)

In [19]:
df.size

191084320

First need to address that the Unamed:0 column and save that as true. 

In [20]:
df.drop("Unnamed: 0", axis=1, inplace=True)

In [21]:
df.head(100)

columns_to_see_classification_error_in = ["segmentsEquipmentDescription", "segmentsAirlineCode", "segmentsAirlineName", "segmentsArrivalAirportCode", "segmentsArrivalTimeEpochSeconds", "segmentsDepartureTimeEpochSeconds", "segmentsCabinCode"]

legId  searchDate  flightDate startingAirport  \
0   79eda9f841e226a1e2121d74211e595c  2022-04-16  2022-04-17             ATL   
1   24c07cf38de8a7a0ed2d7709168042a8  2022-04-16  2022-04-17             ATL   
2   a291a2b5ac19223c3407cfec58d1fc24  2022-04-16  2022-04-17             ATL   
3   d53bcf9e8b6523fa2de9420d7cb22196  2022-04-16  2022-04-17             ATL   
4   5bb138f39804320ece767e747608a50e  2022-04-16  2022-04-17             ATL   
5   8201f75d221c7aa5b78f3b7781166032  2022-04-16  2022-04-17             ATL   
6   a63591f2b4df39d809c4a45a3898e122  2022-04-16  2022-04-17             ATL   
7   79352689384430e23285d3b851fb641e  2022-04-16  2022-04-17             ATL   
8   d9c16c729a1cd13eb08161a9d60b70d8  2022-04-16  2022-04-17             ATL   
9   363081f0b1bbcad5bae624e98b70d23c  2022-04-16  2022-04-17             ATL   
10  ac3e30651b0c6104c432c982975b2ba5  2022-04-16  2022-04-17             ATL   
11  ea05d5dc0cda9644331060d7b837a594  2022-04-16  2022-04-17             ATL   
12  6a30d32e56bbe2be3172e6a49cc80755  2022-04-16  2022-04-17             ATL   
13  c78a14e1a9997ad53554f005585b2f7b  2022-04-16  2022-04-17             ATL   
14  ec36c5799538f25b107ebf1f8c26c605  2022-04-16  2022-04-17             ATL   
15  7cb2e564e0aaf26f7e4e67c6aa5976fd  2022-04-16  2022-04-17             BOS   
16  8e983803b8063ac9afb9587e0e4ae6ba  2022-04-16  2022-04-17             BOS   
17  fdd6daa3c9b3a4dd682287773bf85813  2022-04-16  2022-04-17             BOS   
18  7eefbe958690e01d4c2a25d0d2ebbad9  2022-04-16  2022-04-17             BOS   
19  6be7a3fde436f76934a25338d037cc86  2022-04-16  2022-04-17             BOS   
20  b6de15dddadc49235ef3381d37e04b16  2022-04-16  2022-04-17             BOS   
21  960aae15314fd7a82482c3dabe07e0ba  2022-04-16  2022-04-17             BOS   
22  afd4c60c08d6db335eae1d6f295b9f39  2022-04-16  2022-04-17             BOS   
23  20b4546a9e446522b22b94d214357680  2022-04-16  2022-04-17             BOS   
24  5ece3c09d3663cd221da5ed7b26f16a0  2022-04-16  2022-04-17             BOS   
25  b25d1427118ed12e3637fbbe25709805  2022-04-16  2022-04-17             BOS   
26  85b625309c266b77519d6c6a287bfb7e  2022-04-16  2022-04-17             BOS   
27  fe6ae926f3bc3a72313203f6d57f91ca  2022-04-16  2022-04-17             BOS   
28  1460697a8f8af3c0f1309ba06d492fea  2022-04-16  2022-04-17             BOS   
29  47c1ed628bcd7c659de1ae627de27843  2022-04-16  2022-04-17             BOS   
30  b01e5c47ea2316fae06f84c197706010  2022-04-16  2022-04-17             BOS   
31  f9ed706fc3fde426f5a95f4f82530ae6  2022-04-16  2022-04-17             BOS   
32  1f1f3fadfaa1352c09dc717ab1fc3c7e  2022-04-16  2022-04-17             BOS   
33  4ae2a3e4dd849f46caf38b523d49f22e  2022-04-16  2022-04-17             BOS   
34  98018d5dbaca94f695d0b98ccbe3b68a  2022-04-16  2022-04-17             BOS   
35  7e364b34fd4d25101b3cb4904a21c92d  2022-04-16  2022-04-17             BOS   
36  804c18b7885d5c6f6ec7fc3c6a21da74  2022-04-16  2022-04-17             BOS   
37  f607ec920c820295169ffc85350f24d1  2022-04-16  2022-04-17             BOS   
38  ce8f49affa0a0c5afbe1c8f75c23e5a7  2022-04-16  2022-04-17             BOS   
39  1ef7a9f4cb7aacc698085b3a7f699b6a  2022-04-16  2022-04-17             BOS   
40  dc6efd83dfdd0e17b99bb90d359d3c5c  2022-04-16  2022-04-17             BOS   
41  4b97b6b5690d0298ea18b9d69d993c49  2022-04-16  2022-04-17             BOS   
42  325e035c0f8ec0400230a12481038ae5  2022-04-16  2022-04-17             BOS   
43  7df5521de5a98cc48060165c03291259  2022-04-16  2022-04-17             BOS   
44  d0cc337001ef6092cf96530e2bf75c12  2022-04-16  2022-04-17             BOS   
45  4d83b713def826c7c3684345349529bd  2022-04-16  2022-04-17             BOS   
46  80b2df722618d2f9b7f344cdb00c0a01  2022-04-16  2022-04-17             BOS   
47  2a75e7b99d100eca5cd07c6395c2a5b2  2022-04-16  2022-04-17             BOS   
48  af37cef7a840e30dddf1273d1a49002f  2022-04-16  2022-04-17             BOS   
49  bf7d21670ca9f270edd68dbd402

# Null Handling

In [22]:
print("Finding nulls per column")
df.isnull().any()

Finding nulls per column


legId                                False
searchDate                           False
flightDate                           False
startingAirport                      False
destinationAirport                   False
fareBasisCode                        False
travelDuration                       False
elapsedDays                          False
isBasicEconomy                       False
isRefundable                         False
isNonStop                            False
baseFare                             False
totalFare                            False
seatsRemaining                       False
totalTravelDistance                   True
segmentsDepartureTimeEpochSeconds    False
segmentsDepartureTimeRaw             False
segmentsArrivalTimeEpochSeconds      False
segmentsArrivalTimeRaw               False
segmentsArrivalAirportCode           False
segmentsDepartureAirportCode         False
segmentsAirlineName                  False
segmentsAirlineCode                  False
segmentsEqu

In [59]:
#looking at when `totalTravelDistance` is null
print("Where totalTravelDistance is null")
df_but_null_total_dist = df[
    df["totalTravelDistance"].isnull()
][["totalTravelDistance", "segmentsDistance"]]
display(df_but_null_total_dist.head())


#looking at when `segmentsDistance` is null
print("Where segmentsDistance is null")
df_but_null_segment_dist = df[
    df["segmentsDistance"].isnull()
][["totalTravelDistance", "segmentsDistance"]]
display(df_but_null_segment_dist.head())


#looking at when `segmentsDistanc` and `totalTravelDistance` is null
print("Where segmentsDistance and totalTravelDistance is null")
df_but_null_total_and_segment_dist = df[
    (df["segmentsDistance"].isnull()) & 
    (df["segmentsDistance"].isnull())
][["totalTravelDistance", "segmentsDistance"]]
display(df_but_null_total_and_segment_dist.head())

Where totalTravelDistance is null


Empty DataFrame
Columns: [totalTravelDistance, segmentsDistance]
Index: []

Where segmentsDistance is null


Empty DataFrame
Columns: [totalTravelDistance, segmentsDistance]
Index: []

Where segmentsDistance and totalTravelDistance is null


Empty DataFrame
Columns: [totalTravelDistance, segmentsDistance]
Index: []

In [23]:
print("Observing null columns")
df[["totalTravelDistance", "segmentsEquipmentDescription", "segmentsDistance"]].head()

Observing null columns


totalTravelDistance segmentsEquipmentDescription segmentsDistance
0                947.0                          NaN              947
1                947.0                          NaN              947
2               1675.0                ||Embraer 190         947||728
3               1247.0     Embraer 190||Embraer 190         762||485
4                762.0                  Embraer 190              762

In [24]:
null_col_sums = df.isna().sum()
display(null_col_sums)
null_numerators = [x for x in null_col_sums if x!= 0]

denominator = df.shape[0] #we are using the shape tuple to pull in the total rows to accurately understand how much is missing on a grand perspective

for null_numerator in null_numerators:
    print((null_numerator/denominator)*100)

legId                                    0
searchDate                               0
flightDate                               0
startingAirport                          0
destinationAirport                       0
fareBasisCode                            0
travelDuration                           0
elapsedDays                              0
isBasicEconomy                           0
isRefundable                             0
isNonStop                                0
baseFare                                 0
totalFare                                0
seatsRemaining                           0
totalTravelDistance                  56436
segmentsDepartureTimeEpochSeconds        0
segmentsDepartureTimeRaw                 0
segmentsArrivalTimeEpochSeconds          0
segmentsArrivalTimeRaw                   0
segmentsArrivalAirportCode               0
segmentsDepartureAirportCode             0
segmentsAirlineName                      0
segmentsAirlineCode                      0
segmentsEqu

0.8269689527638898
0.7684586574136486
0.2852834811354485


With the JetBlue data subset only showing around 0.82% of nulls existent (total elements...not rows), the executive decision that comes to mind is to remove the nulls. However, one thing to still do is to count the unique pairings of these elements to understand exactly what we are removing. 

In [25]:
#find a way to filter all unique pairings of T/F...
#does this warrant some feature engineering?

df["totalTravelDistance_null_existence"] =  df["totalTravelDistance"].isnull()
df["segmentsDistance_null_existence"] = df["segmentsDistance"].isnull()
df["segmentsEquipmentDescription_null_existence"] = df["segmentsEquipmentDescription"].isnull()

df.head()

legId  searchDate  flightDate startingAirport  \
0  79eda9f841e226a1e2121d74211e595c  2022-04-16  2022-04-17             ATL   
1  24c07cf38de8a7a0ed2d7709168042a8  2022-04-16  2022-04-17             ATL   
2  a291a2b5ac19223c3407cfec58d1fc24  2022-04-16  2022-04-17             ATL   
3  d53bcf9e8b6523fa2de9420d7cb22196  2022-04-16  2022-04-17             ATL   
4  5bb138f39804320ece767e747608a50e  2022-04-16  2022-04-17             ATL   

  destinationAirport fareBasisCode travelDuration  elapsedDays  \
0                BOS      VH0AUEL1        PT2H38M            0   
1                BOS      QH0KUEL1        PT2H44M            0   
2                CLT      VH0AUEL1        PT8H48M            0   
3                DTW      RH0AUEL1         PT5H8M            0   
4                EWR      VC0QBEL1         PT2H9M            0   

   isBasicEconomy  isRefundable  isNonStop  baseFare  totalFare  \
0           False         False       True    217.67     248.60   
1           False         False       True    316.73     355.09   
2           False         False      False    482.79     542.60   
3           False         False      False    269.78     313.61   
4           False         False       True    171.16     198.60   

   seatsRemaining  totalTravelDistance segmentsDepartureTimeEpochSeconds  \
0               7                947.0                        1650206700   
1               7                947.0                        1650233220   
2               7               1675.0            1650206700||1650229500   
3               7               1247.0            1650235800||1650246600   
4               7                762.0                        1650195000   

                            segmentsDepartureTimeRaw  \
0                      2022-04-17T10:45:00.000-04:00   
1                      2022-04-17T18:07:00.000-04:00   
2  2022-04-17T10:45:00.000-04:00||2022-04-17T17:0...   
3  2022-04-17T18:50:00.000-04:00||2022-04-17T21:5...   
4                      2022-04-17T07:30:00.000-04:00   

  segmentsArrivalTimeEpochSeconds  \
0                      1650216180   
1                      1650243060   
2          1650216180||1650238380   
3          1650244380||1650254280   
4                      1650202740   

                              segmentsArrivalTimeRaw  \
0                      2022-04-17T13:23:00.000-04:00   
1                      2022-04-17T20:51:00.000-04:00   
2  2022-04-17T13:23:00.000-04:00||2022-04-17T19:3...   
3  2022-04-17T21:13:00.000-04:00||2022-04-17T23:5...   
4                      2022-04-17T09:39:00.000-04:00   

  segmentsArrivalAirportCode segmentsDepartureAirportCode  \
0                        BOS                          ATL   
1                        BOS                          ATL   
2                   BOS||CLT                     ATL||BOS   
3                   JFK||DTW                     ATL||JFK   
4                        EWR                          ATL   

                segmentsAirlineName segmentsAirlineCode  \
0                   JetBlue Airways                  B6   
1                   JetBlue Airways                  B6   
2  JetBlue Airways||JetBlue Airways              B6||B6   
3  JetBlue Airways||JetBlue Airways              B6||B6   
4                   JetBlue Airways                  B6   

  segmentsEquipmentDescription segmentsDurationInSeconds segmentsDistance  \
0                          NaN                      9480              947   
1                          NaN                      9840              947   
2                ||Embraer 190                9480||8880         947||728   
3     Embraer 190||Embraer 190                8580||7680         762||485   
4                  Embraer 190                      7740              762   

  segmentsCabinCode  totalTravelDistance_null_existence  \
0             coach                               False   
1             coach                               False   
2      coach||coach                    

In [26]:
print(df.groupby(["totalTravelDistance_null_existence", "segmentsDistance_null_existence", "segmentsEquipmentDescription_null_existence"]).size())

print(f'The length of the df null existence series is: {len(df.groupby(["totalTravelDistance_null_existence", "segmentsDistance_null_existence", "segmentsEquipmentDescription_null_existence"]).size())}')

totalTravelDistance_null_existence  segmentsDistance_null_existence  segmentsEquipmentDescription_null_existence
False                               False                            False                                          6715947
                                                                     True                                             52057
True                                False                            False                                            36967
                                    True                             False                                            19083
                                                                     True                                               386
dtype: int64
The length of the df null existence series is: 5


From the top down, we have a series of the values where we grouped the column pairings of the True/False null existence features. The length is 5 and we're going to only select relevant features here. 

In [27]:
len(df.groupby(["totalTravelDistance_null_existence", "segmentsDistance_null_existence", "segmentsEquipmentDescription_null_existence"]).size())

5

In [28]:
df.groupby(["totalTravelDistance_null_existence", "segmentsDistance_null_existence", "segmentsEquipmentDescription_null_existence"]).size().sum()

6824440

In [35]:
distinct_null_counter_array = df.groupby(["totalTravelDistance_null_existence", "segmentsDistance_null_existence", "segmentsEquipmentDescription_null_existence"]).size()

print(f"""
The nulls we are concerned with: {distinct_null_counter_array[1]} + {distinct_null_counter_array[2]} + {distinct_null_counter_array[4]}
""")

result = distinct_null_counter_array[1] + distinct_null_counter_array[2] + distinct_null_counter_array[4]

print(f"This is the total amount of loss: {result}")

print(f"The percentage of total nulls compounded by distinction is: {result / df.groupby(['totalTravelDistance_null_existence', 'segmentsDistance_null_existence', 'segmentsEquipmentDescription_null_existence']).size().sum()}")


The nulls we are concerned with: 52057 + 36967 + 386

This is the total amount of loss: 89410
The percentage of total nulls compounded by distinction is: 0.0131014412904209


/var/folders/gf/trjkr8p14xbfkbwsfn2w5_kw0000gn/T/ipykernel_2171/3343332512.py:4: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  The nulls we are concerned with: {distinct_null_counter_array[1]} + {distinct_null_counter_array[2]} + {distinct_null_counter_array[4]}
/var/folders/gf/trjkr8p14xbfkbwsfn2w5_kw0000gn/T/ipykernel_2171/3343332512.py:7: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  result = distinct_null_counter_array[1] + distinct_null_counter_array[2] + distinct_null_counter_array[4]


So we found the true data loss due to nulls and the result is 0.01% of our data to be removed (summing to a total of 89,410 rows out of our 6.8M long dataframe. Let's remove the total code recognized nulls.)

In [38]:
df.dropna(inplace=True)

In [41]:
df.shape

(6715947, 30)

# Error Checking

**Cleaning Poor Labeling (Equipment)**

- I saw some places where `"||"` exists for either segments distance or equipment. 
    - To see such, I should look at the distinct values of these dataframes. 

I saw some places where `"||"` exists for either segments distance or equipment. 

To see such, I should look at the distinct values of these dataframes. 

In [63]:
columns_to_see_classification_error_in = ["segmentsEquipmentDescription", "segmentsAirlineCode", "segmentsAirlineName", "segmentsArrivalAirportCode", "segmentsArrivalTimeEpochSeconds", "segmentsDepartureTimeEpochSeconds", "segmentsCabinCode"]

sorted(df[columns_to_see_classification_error_in[0]].unique())

['AIRBUS INDUSTRIE A321 SHARKLETS',
 'AIRBUS INDUSTRIE A321 SHARKLETS||',
 'AIRBUS INDUSTRIE A321 SHARKLETS||AIRBUS INDUSTRIE A321 SHARKLETS',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Airbus A319',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Airbus A319-321',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Airbus A319-321||Airbus A320',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Airbus A320',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Airbus A321',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Boeing 737 MAX 8',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Boeing 737-800',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Boeing 777-200',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Embraer 175',
 'AIRBUS INDUSTRIE A321 SHARKLETS||Embraer 190',
 'Airbus A319',
 'Airbus A319-321',
 'Airbus A319-321||',
 'Airbus A319-321||AIRBUS INDUSTRIE A321 SHARKLETS',
 'Airbus A319-321||Airbus A319',
 'Airbus A319-321||Airbus A320',
 'Airbus A319-321||Airbus A320||',
 'Airbus A319-321||Airbus A320||Airbus A320',
 'Airbus A319-321||Airbus A320||Embraer 190',
 'Airbus A319-321||Airbus A321'

In [68]:
df.groupby(columns_to_see_classification_error_in[0]).size().sort_values(ascending=False)

segmentsEquipmentDescription
Boeing 737-800                                                                                                                    1152884
Airbus A319-321                                                                                                                    490818
Embraer 190                                                                                                                        482274
Embraer 190||Embraer 190                                                                                                           428326
Airbus A320                                                                                                                        396667
Airbus A320||Airbus A320                                                                                                           318251
Airbus A321                                                                                                                        294106
Airbu

`regex` might be necessary here to identify those areas where we notice some problematic strings. However, identifying such can be done by first filtering down to anything that contains at least `|`. 

In [77]:
df[
    df[columns_to_see_classification_error_in[0]].str.contains("|||", case=False, regex=False) 
].groupby(columns_to_see_classification_error_in[0]).size().sort_values(ascending=False) #regex is false because | is naturally treated as an "OR" command in python

segmentsEquipmentDescription
Airbus A320||||Embraer 190        585
Embraer 190||||Airbus A319-321    237
Airbus A320||||Airbus A320        202
Embraer 190||||Airbus A320        131
Airbus A320||||                    40
Embraer 190||||Embraer 190         13
Airbus A320||||Airbus A321         10
||||Airbus A320                     3
Airbus A321||||Airbus A320          1
Embraer 190||||Boeing 737-800       1
||||Embraer 190                     1
dtype: int64

In [76]:
len(df[
    df[columns_to_see_classification_error_in[0]].str.contains("|", case=False, regex=False) 
].groupby(columns_to_see_classification_error_in[0]).size().sort_values(ascending=False)) #regex is false because | is naturally treated as an "OR" command in python

255

In [79]:
df[
    df[columns_to_see_classification_error_in[0]].str.contains("|||", case=False, regex=False) 
]

legId  searchDate  flightDate  \
2415995  c4e9eb3627ed22dd0b8b58e2cac14681  2022-06-25  2022-08-04   
2415996  152cbd60711e9ec28362a6b2dd492ece  2022-06-25  2022-08-04   
2415997  f94ad2f45f9f92886e17e2cd941fad67  2022-06-25  2022-08-04   
2415999  ad8a8023b66e87971b3ccaac1396003f  2022-06-25  2022-08-04   
2416000  b08919497386c2d5ebada933b7980cd7  2022-06-25  2022-08-04   
...                                   ...         ...         ...   
6799173  ba5e50469b87ac7d38d9b12f82071748  2022-10-05  2022-10-18   
6810554  4719a18ae3a2821b8d1b4ed9409c32f4  2022-10-05  2022-10-29   
6812442  b2f13a6573609e8d72a0c276b613ecd1  2022-10-05  2022-10-31   
6820610  44e53a36361b688562b4e01d8c64a841  2022-10-05  2022-11-08   
6823751  52235591facfb25c5df76c3bba28b9c0  2022-10-05  2022-11-11   

        startingAirport destinationAirport fareBasisCode travelDuration  \
2415995             PHL                LAX      PL2ABEL1       PT16H52M   
2415996             PHL                LAX      PL2ABEL1        PT18H8M   
2415997             PHL                LAX      PL2ABEL1       PT19H14M   
2415999             PHL                LAX      OL4AUEL1       PT12H47M   
2416000             PHL                LAX      OL4AUEL1        PT14H3M   
...                 ...                ...           ...            ...   
6799173             LAX                EWR      PI7QUSL1       PT14H59M   
6810554             LAX                EWR      PI7QUSL1       PT15H22M   
6812442             JFK                DTW      SI4QUEL1        PT15H6M   
6820610             JFK                DTW      PI4QUSL1        PT15H2M   
6823751             JFK                DTW      PI4QUEL1        PT15H2M   

         elapsedDays  isBasicEconomy  isRefundable  isNonStop  baseFare  \
2415995            0           False         False      False    346.98   
2415996            1           False         False      False    346.98   
2415997            1           False         False      False    346.98   
2415999            0           False         False      False    374.88   
2416000            1           False         False      False    374.88   
...              ...             ...           ...        ...       ...   
6799173            1           False         False      False    218.61   
6810554            1           False         False      False    101.40   
6812442            1           False         False      False    127.45   
6820610            1           False         False      False    108.85   
6823751            1           False         False      False    118.15   

         totalFare  seatsRemaining  totalTravelDistance  \
2415995     406.70               7               2923.0   
2415996     406.70               7               2923.0   
2415997     406.70               7               2923.0   
2415999     431.10               7               2923.0   
2416000     431.10               7               2923.0   
...            ...             ...                  ...   
6799173     268.71               7               2805.0   
6810554     142.71               7               2805.0   
6812442     170.71               7               1664.0   
6820610     150.71               7               1664.0   
6823751     160.71               7               1664.0   

          segmentsDepartureTimeEpochSeconds  \
2415995  1659620700||1659650940||1659658500   
2415996  1659620700||1659650940||1659663600   
2415997  1659620700||1659650940||1659667500   
2415999  1659635400||1659650940||1659658500   
2416000  1659635400||1659650940||1659663600   
...                                     ...   
6799173  1666136400||1666161000||1666184400   
6810554  1667086800||1667111400||1667136600   
6812442  1667267940||1667297340||1667314140   
6820610  1667962740||1667992140||1668008700   
6823751  1668221940||1668251340||1668267900   

                                  segmentsDepartureTimeRaw  \
2415995  2022-08-04T09:45:00.000-04:00||2022-08-04T18:0...   
2415996

There is more error with the `segmentsEquipmentDescription`. When referencing some trips with multiple legs (seeing it displayed in the `segmentsAirlineCode` column), there are multiple leg trips. Great. However, the labeling of the equipment used to travel is incorrect. For instance, `legId` `c4e9eb3627ed22dd0b8b58e2cac14681` shows that there are 3 legs to the trip, but the equipment is separated by one delimiter of `||||`. <-- side note, this is interesting in count because it showcases enough `||` to house one more element. 

Anyway...the aircraft shown here is: `Embraer 190||||Airbus A319-321`. The Airbus A319-321 is an aircraft that does not exist. What might be implied is that there are two aircrafts being spoken of on this Airbus comment. I believe the legs correspond to two Airbus aircraft types: A319 and A321. 

How is this to be addressed and cleaned? 

I may have to drop this altogether. Or otherwise, consider removing multi-leg trips. Or consider bringing in another dataset to verify the integrity of these aircraft. Or consider create a rule that states to only look at any thing that has `||` only in length + direct routes. Or I remove anything that has a `-` in its aircraft name. This last suggesstion isn't one I would be proud of. 

**

# Data Integrity Check (outlier detection)

Make sure we aren't getting an outrageous values. 

**Unrealistic Values**

- check dates aren't outrageous
- check that mileage isn't outrageous
- check that seat numbers aren't outrageous
- 

**Outlier Detection**

# Feature Transformation

Let's address a conversion of time and maybe considering an additional feature that best represents time. S

Can we verify the seconds from one column actually equates to the dates in another. 

Do we need to clean one feature of legs into two or 3 features? Using additional features to understand important legs of trip. 

**Removing Duplicated Features**

- test to remove Airline Code and Airline Name
    - need to assert that there are a limited amount of distinct values (if so, then it means the feature is not that important as we aren't using it compare against other airlines).

In [80]:
##airline code uniqueness

sorted(df[columns_to_see_classification_error_in[1]].unique())

['9K||9K||9K||B6',
 '9K||9K||B6',
 '9K||B6',
 '9K||B6||B6',
 'AA||B6',
 'B6',
 'B6||3M||B6',
 'B6||9K',
 'B6||9K||9K',
 'B6||9K||B6',
 'B6||AA',
 'B6||B6',
 'B6||B6||9K',
 'B6||B6||B6']

In [81]:
df.groupby(columns_to_see_classification_error_in[1]).size().sort_values(ascending=False)

segmentsAirlineCode
B6                3506867
B6||B6            3092858
B6||B6||B6         108877
B6||9K||9K           2142
9K||B6               1842
B6||9K               1353
B6||9K||B6           1035
AA||B6                388
9K||9K||B6            249
B6||B6||9K            126
9K||B6||B6            116
B6||3M||B6             80
B6||AA                 13
9K||9K||9K||B6          1
dtype: int64

In [82]:
##airline

sorted(df[columns_to_see_classification_error_in[2]].unique())

['American Airlines||JetBlue Airways',
 'Cape Air||Cape Air||Cape Air||JetBlue Airways',
 'Cape Air||Cape Air||JetBlue Airways',
 'Cape Air||JetBlue Airways',
 'Cape Air||JetBlue Airways||JetBlue Airways',
 'JetBlue Airways',
 'JetBlue Airways||American Airlines',
 'JetBlue Airways||Cape Air',
 'JetBlue Airways||Cape Air||Cape Air',
 'JetBlue Airways||Cape Air||JetBlue Airways',
 'JetBlue Airways||JetBlue Airways',
 'JetBlue Airways||JetBlue Airways||Cape Air',
 'JetBlue Airways||JetBlue Airways||JetBlue Airways',
 'JetBlue Airways||Silver Airways||JetBlue Airways']

In [83]:
df.groupby(columns_to_see_classification_error_in[2]).size().sort_values(ascending=False)

segmentsAirlineName
JetBlue Airways                                      3506867
JetBlue Airways||JetBlue Airways                     3092858
JetBlue Airways||JetBlue Airways||JetBlue Airways     108877
JetBlue Airways||Cape Air||Cape Air                     2142
Cape Air||JetBlue Airways                               1842
JetBlue Airways||Cape Air                               1353
JetBlue Airways||Cape Air||JetBlue Airways              1035
American Airlines||JetBlue Airways                       388
Cape Air||Cape Air||JetBlue Airways                      249
JetBlue Airways||JetBlue Airways||Cape Air               126
Cape Air||JetBlue Airways||JetBlue Airways               116
JetBlue Airways||Silver Airways||JetBlue Airways          80
JetBlue Airways||American Airlines                        13
Cape Air||Cape Air||Cape Air||JetBlue Airways              1
dtype: int64

Additional reduction in some of these may be necessary. I propose first ridding all the rows that have a different airline labeled that isn't JetBlue. These airlines included in routes may sway price, but they complicate cleaning for a future model. Furthermore, they are less significant thatn the rest of the data which may also introduce some bias in the modeling. I'm going to count how many unique value instances comprise of only JetBlue flights and then those that are mixed. We want to remove the mixed version to make a strong JetBlue model. 

**KPI Creation**

- dollar/mile or mile/dollar
    - mile/dollar is easier to understand; dollar/mile is less intuitive as we're trying to use pennies swaying to measure vs mile swaying
- averages and medians for classification feature
- standard deviation of stuff

# Feature Reduction

Consider if `segmentsAirlineCode` is really necessary. It seems to only contain `B6` for airline codes which is a common Jet Blue indicator. 